File: C++/Step19-Binary Search/1654.cpp
//
//  1654.cpp
//  Step19-Binary Search
//
//  Created by 김예빈 on 2019. 9. 28..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
#include <cmath>
using namespace std;

int main(int argc, const char * argv[]) {
    int k, n, result;
    long long max = 0, mid, left = 0, right = pow(2, 31) - 1;
    
    cin >> k >> n;
    
    int arr[k];
    
    for(int i = 0; i < k; i++)
        cin >> arr[i];
    
    while(left <= right) {
        result = 0;
        mid = (left + right) / 2;
        
        for(int j = 0; j < k; j++)
            result = result + arr[j] / mid;
        
        if(result >= n) {
            left = mid + 1;
            
            if(mid > max)
                max = mid;
        } else if(result < n)
            right = mid-1;
    }
    
    cout << max;
    
    return 0;
}

File: BaekJoon Online Judge/Swift/Step20-Binary Search/1654.swift
//
//  1654.swift
//  Step20-Binary Search
//
//  Created by Yebin Kim on 2020/01/05.
//  Copyright © 2020 김예빈. All rights reserved.
//

import Foundation

let input = readLine()!.split(separator: " ").map{ UInt($0)! }
let K = input[0]
let N = input[1]
var answer: UInt = 0

let cables = (0 ..< K).reduce(into: [UInt]()) { (sum, _) in
    sum.append(UInt(readLine()!)!)
}.sorted(by: >)

var left: UInt = 1
var right: UInt = cables.first!

while left <= right {
    let mid: UInt = (left + right) / 2
    var total: UInt = 0
    
    for cable in cables {
        let quot: UInt = cable / mid
        if quot > 0 {
            total += quot
        } else {
            break
        }
    }
    
    if total >= N {
        answer = max(mid, answer)
        left = mid + 1
    } else {
        right = mid - 1
    }
    
}

print(answer)
